### PR TITLE
Markdown ⬇️

### LINKED ISSUE #73: Add Markdown
<a href="https://github.com/shawnajean"></img></a> **Issue by [shawnajean](https://github.com/shawnajean)**
_Friday Jun 24, 2016 at 17:57 GMT_

---

This is a pretty common request, though I'm not seeing it on the feature request site.

User requested to be notified when the feature is added: 2712422-t

### PR DESCRIPTION
**Pull request by roundhill**
I apologize in advance for the size of this PR 😬 

Let's add markdown support! This PR adds some great new features to the editor:

* The ability to toggle markdown for a note in the `info` menu.
* A 'live' markdown editor using a modified [Notepad](https://github.com/ruddfawcett/Notepad) library that works with our theming system. It parses markdown and applies `NSAttributedStrings` as you're typing. Performance seems good to me but we should stress test it! If it works well I'd also like to add it to the iOS app.
* A new 'preview' button in the toolbar that will load the content in a `WKWebView`. The web view will only allow http, https or mailto links to be clicked. The preview button should only show if the note has markdown enabled.

Demo:
jan-23-2018 12-13-14

Fixes #73

### CODE REVIEW

File: Simplenote/NoteEditorViewController.m [hunk 1]
```diff
@@ -1160,4 +1147,25 @@ - (NSPopover *)newPopoverWithContentViewController:(NSViewController *)viewContr
...
+#pragma mark - WKNavigationDelegate
+
+- (void)webView:(WKWebView *)webView decidePolicyForNavigationAction:(WKNavigationAction *)navigationAction decisionHandler:(void (^)(WKNavigationActionPolicy))decisionHandler {
+    if (navigationAction.navigationType == WKNavigationTypeLinkActivated) {
+        NSURL *linkUrl = navigationAction.request.URL;
+        if ([self urlSchemeIsAllowed:linkUrl.scheme]) {
+            [[NSWorkspace sharedWorkspace] openURL:linkUrl];
+        }
+        
+        decisionHandler(WKNavigationActionPolicyCancel);
```
Comment (jleandroperez) [inline Simplenote/NoteEditorViewController.m]: We should probably bail out once the handler is called. Can we add a `return` here?

File: Simplenote/NoteEditorViewController.m [hunk 2]
```diff
@@ -97,7 +98,7 @@ - (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibB
     self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
     
     if (self) {
```
Comment (frosty) [inline Simplenote/NoteEditorViewController.m]: I guess you could remove all of this `if` :)

File: Simplenote/NoteEditorViewController.m [hunk 3]
```diff
@@ -588,9 +588,8 @@ - (void)menuWillOpen:(NSMenu *)menu
     
     [self updateCounts];
     pinnedItem.state = [self selectedNotesPinned] ? NSOnState : NSOffState;
+    markdownItem.state = [self selectedNotesMarkdowned] ? NSOnState : NSOffState;
     [collaborateItem setEnabled:numSelectedNotes == 1];
-    [publishItem setEnabled:numSelectedNotes == 1];
```
Comment (frosty) [inline Simplenote/NoteEditorViewController.m]: Should these lines have been removed? The publish icon is still enabled now if I select multiple notes, but doesn't do anything if I click it. And the history popup shows 'version: null' if I click it with multiple notes enabled.

File: Simplenote/NoteEditorViewController.m [hunk 4]
```diff
@@ -1160,4 +1147,25 @@ - (NSPopover *)newPopoverWithContentViewController:(NSViewController *)viewContr
...
+        [scheme isEqualToString:@"https"] ||
+        [scheme isEqualToString:@"mailto"];
+}
+
+#pragma mark - WKNavigationDelegate
+
+- (void)webView:(WKWebView *)webView decidePolicyForNavigationAction:(WKNavigationAction *)navigationAction decisionHandler:(void (^)(WKNavigationActionPolicy))decisionHandler {
+    if (navigationAction.navigationType == WKNavigationTypeLinkActivated) {
+        NSURL *linkUrl = navigationAction.request.URL;
+        if ([self urlSchemeIsAllowed:linkUrl.scheme]) {
```
Comment (frosty) [inline Simplenote/NoteEditorViewController.m]: It'd be nice if we could allow anchor links, too (within the current document) – useful for tables of content.

File: Simplenote/NoteEditorViewController.m (hunk 3 above)
Comment (roundhill) [reply Simplenote/NoteEditorViewController.m]: Nope! Not sure how that happened :) Nice catch.

File: Simplenote/NoteEditorViewController.m (hunk 4 above)
Comment (roundhill) [reply Simplenote/NoteEditorViewController.m]: > It'd be nice if we could allow anchor links, too

Darn, I couldn't get this to work. When you do `[click me](#myanchor)` macOS thinks the link is a file url, which is not so good security-wise.

**Comment by roundhill** [comment]
@frosty / @jleandroperez Thank you for the feedback! I've fixed the issues you've reported, please have another look! I've also disabled the `SPTextLinkifier` in order to fix a few bugs with opening links. This may impact performance with extremely large notes though, so please give that a test.

**Comment by roundhill** [comment]
> spellcheck highlighting got messed up

Yikes! Will see if I can reproduce that. Do you think maybe it was trying to spell check the content of the markdown preview?

**Comment by frosty** [comment]
It doesn't seem related to the preview. I've just been able to get it to happen a few more times simply by clicking rapidly up through 8 or 9 notes. It only seems to happen when I end on this one specific note though, so maybe it's something in the content there.

notes

The content is the Markdown syntax specification: https://daringfireball.net/projects/markdown/syntax.text

**Comment by frosty** [comment]
Interestingly it's not just a graphical glitch - the spelling suggestions are 'correct' for the highlighted areas of text:

screen shot 2018-02-08 at 10 39 07

screen shot 2018-02-08 at 10 38 48

**Comment by roundhill** [comment]
Thanks all for the reviews!

**Pull request merged**